src/linear_algebra/matrix/diag.lean

```diff
@@ -0,0 +1,226 @@
+  Aᵀ.is_diag :=
+by { intros i j h, simp [ha h.symm] }
+
+@[simp] lemma is_diag.conj_transpose
+  [semiring α] [star_ring α] {A : matrix n n α} (ha : A.is_diag) :
+  Aᴴ.is_diag :=
+by { intros i j h, simp [ha h.symm] }
+
+@[simp] lemma is_diag.minor [has_zero α]
+  {A : matrix n n α} (ha : A.is_diag) {f : m → n} (hf : injective f) :
+  (A.minor f f).is_diag :=
+by { intros i j h, specialize @hf i j, simp [ha (mt hf h)] }
+
+/-- `(A ⊗ B).is_diag` if both `A` and `B` are diagonal. -/
+@[simp] lemma is_diag.kronecker [mul_zero_class α]
+  {A : matrix m m α} {B : matrix n n α} (hA: A.is_diag) (hB: B.is_diag) :
+  (A ⊗ₖ B).is_diag :=
+begin
+  rintros ⟨a, b⟩ ⟨c, d⟩ h,
+  dsimp [kronecker_apply],
+  by_cases hac: a = c,
+  { have hbd: b ≠ d, {tidy}, simp [hB hbd] },
+  { simp [hA hac] },
+end
+
+@[simp] lemma is_diag.diagonal [has_zero α] [decidable_eq n] (d : n → α) :
+  (diagonal d).is_diag :=
+by { intros i j h, simp [h] }
+
+lemma is_diag.is_symm [has_zero α] {A : matrix n n α} (h : A.is_diag) :
+  A.is_symm :=
+begin
+  ext,
+  by_cases g : i = j, {rw g},
+  simp [h g, h (ne.symm g)],
+end
+
+/-- The block matrix `A.from_blocks 0 0 D` is diagonal if  `A` and `D` are diagonal. -/
+lemma is_diag.from_blocks [has_zero α]
+  {A : matrix m m α} {D : matrix n n α}
+  (ha : A.is_diag) (hd : D.is_diag) :
+  (A.from_blocks 0 0 D).is_diag :=
+begin
+  rintros (i | i) (j | j) hij,
+  { have h' := ne_of_apply_ne (λ (i : m), sum.inl i) hij, simp [ha h'] },
+  { simp * at * },
+  { simp * at * },
+  { have h' := ne_of_apply_ne (λ (i : n), sum.inr i) hij, simp [hd h'] },
+end
+
+/-- This is the `iff` version of `matrix.is_diag.from_blocks`. -/
+lemma is_diag_from_blocks_iff [has_zero α]
+  {A : matrix m m α} {B : matrix m n α} {C : matrix n m α} {D : matrix n n α} :
+  (A.from_blocks B C D).is_diag ↔ A.is_diag ∧ B = 0 ∧ C = 0 ∧ D.is_diag :=
+begin
+  split,
+  { intros h,
+    repeat {split};
+    try {intros i j hij <|> ext i j},
+    { have g : sum.inl i ≠ sum.inl j, {simp [hij]},
+      specialize h g,
+      simp [from_blocks, *] at * },
+    { have g : sum.inl i ≠ sum.inr j, {simp},
+      specialize h g,
+      simp [from_blocks, *] at * },
+    { have g : sum.inr i ≠ sum.inl j, {simp},
+      specialize h g,
+      simp [from_blocks, *] at * },
+    { have g : sum.inr i ≠ sum.inr j, {simp [hij]},
+      specialize h g,
+      simp [from_blocks, *] at * } },
+  { rintros ⟨ha, hb, hc, hd⟩,
+    convert is_diag.from_blocks ha hd }
+end
+
+/-- A symmetric block matrix `A.from_blocks B C D` is diagonal
+    if  `A` and `D` are diagonal and `B` is `0`. -/
+lemma is_diag.from_blocks_of_is_symm [has_zero α]
+  {A : matrix m m α} {B : matrix m n α} {C : matrix n m α} {D : matrix n n α}
+  (h : (A.from_blocks B C D).is_symm) (ha : A.is_diag) (hb : B = 0) (hd : D.is_diag) :
+  (A.from_blocks B C D).is_diag :=
+begin
+  convert ha.from_blocks hd,
+  obtain ⟨g1, g2, g3, g4⟩ := is_symm_from_blocks_iff.1 h,
+  simp* at *
+end
```

Comment:
I would _maybe_ eliminate `B` from this in the same way as above:
```suggestion
/-- A symmetric block matrix `A.from_blocks B C D` is diagonal
    if  `A` and `D` are diagonal and `B` is `0`. -/
lemma is_diag.from_blocks_of_is_symm [has_zero α]
  {A : matrix m m α} {C : matrix n m α} {D : matrix n n α}
  (h : (A.from_blocks 0 C D).is_symm) (ha : A.is_diag) (hd : D.is_diag) :
  (A.from_blocks 0 C D).is_diag :=
begin
  rw ←(is_symm_from_blocks_iff.1 h).2.1,
  exact ha.from_blocks hd,
end
```
However, I'd also maybe argue this lemma just isn't useful in its current form anyway, and that it should just be removed; `is_symm_from_blocks_iff` and `is_diag_from_blocks_iff` are much more general.

Reply:
I think this lemma is useful when we know the given block matrix is symmetric. This lemma uses both the two lemmas you mentioned as a generalization.